documentation/sphinx/source/client-testing.rst

```diff
@@ -373,3 +373,296 @@ with the ``multitest`` role:
+  depending on the selected API version. The API changes are correctly adapted. 
+- Client API behaves correctly in case of cluster upgrades. Database and transaction state is correctly
+  migrated to the upgraded connections. Pending operations are canceled and successfully retried on the
+  upgraded connections.
+
+Implementing a Workload
+=======================
+
+Each workload is declared as a direct or indirect subclass of ``WorkloadBase`` implementing a constructor
+with ``WorkloadConfig`` as a parameter and the method ``start()``, which defines the entry point of the
+workload. 
+
+``WorkloadBase`` provides a set of methods that serve as building blocks for implementation of a workload:
+
+.. function:: execTransaction(start, cont, failOnError = true)
+
+   creates and executes an FDB transaction. Here ``start`` is a function that takes a transaction context
+   as parameter and implements the starting point of the transaction, and ``cont`` is a function implementing
+   a continuation to be executed after finishing the transaction execution. Transactions are automatically
+   retried on retryable errors. In case of a fatal error, the entire workload is considered as failed
+   unless ``failOnError`` is set to ``false``.
+
+.. function:: schedule(task)
+
+   schedules a task for asynchronous execution. It is usually used in the continuations to schedule 
+   the next step of the workload.
+
+.. function:: info(msg) 
+              error(msg) 
+              
+   are used for logging a message with a tag identifying the workload. Issuing an error message marks
+   the workload as failed.
+
+The transaction context provides methods for implementation of the transaction logics:
+
+.. function:: tx()
+   
+   the reference to the FDB transaction object
+
+.. function:: continueAfter(future, cont, retryOnError = true)
+   
+   set a continuation to be executed when the future is ready. The ``retryOnError`` flag controls whether
+   the transaction should be automatically retried in case the future results in an error.
+
+.. function:: continueAfterAll(futures, cont)
+   
+   takes a vector of futures and sets a continuation to be executed when all of the futures get ready.
+   The transaction is retried if at least one of the futures results in an error. This method is useful 
+   for handling multiple concurrent reads.
+
+.. function:: commit() 
+   
+   commit and finish the transaction. If the commit is successful, the execution proceeds to the
+   continuation of ``execTransaction()``, otherwise the transaction is retried. 
+
+.. function:: done() 
+   
+   finish the transaction without committing. This method should be used to finish read transactions. 
+   The transaction gets destroyed and execution proceeds to the continuation of ``execTransaction()``.
+
+.. function:: onError(err) 
+   
+   retry transaction on an error. This method is typically used in the continuation of continueAfter
```

Comment:
This is a bit confusing. AFAIK FDB C API's `fdb_transaction_on_error()` by itself doesn't retry the transaction. It resets the transaction state based on its retriability and inserts a back-off the client API deems necessary.
If I understand correctly, API tester's `onError()` simply leads to the next iteration of the workload (by calling `start()`, unless the error is terminal) which may not be the same exact transaction (e.g. in the `setAndGet()` example, the key and value would be different). Then it shouldn't be called a "retry".

Reply:
Talking in precise terms you are right. Whether it is a retry of the same transaction or not indeed depends on the implementation of the start function, but the same is in case of the typical retry loops of FDB transactions. The name onError suggests that it is an error handler. Yes, fdb_transaction_on_error() does less, because it still requires coding the retry loop around it.